batch/batch/driver/pool.py

```diff
@@ -209,6 +225,27 @@ async def create_instance(self, cores=None, max_idle_time_msecs=None):
             job_private=False,
         )
 
+    async def create_instances_from_ready_cores(self, ready_cores_mcpu):
+        n_live_instances = self.n_instances_by_state['pending'] + self.n_instances_by_state['active']
+
+        instances_needed = (ready_cores_mcpu - self.live_free_cores_mcpu + (self.worker_cores * 1000) - 1) // (
+            self.worker_cores * 1000
+        )
+        instances_needed = min(
+            instances_needed,
+            self.max_live_instances - n_live_instances,
+            self.max_instances - self.n_instances,
+            # 20 queries/s; our GCE long-run quota
+            300,
+            # n * 16 cores / 15s = excess_scheduling_rate/s = 10/s => n ~= 10
+            10,
+        )
+
+        if instances_needed > 0:
+            log.info(f'creating {instances_needed} new instances')
+            # parallelism will be bounded by thread pool
+            await asyncio.gather(*[self.create_instance() for _ in range(instances_needed)])
+
     async def create_instances(self):
         ready_cores = await self.db.select_and_fetchone(
```

Comment:
I'm inclined to hit the database once rather than twice. The per_user query is sufficient, we can compute the sum over all users in Python. Also, this SQL expression uses a LOCK IN SHARE MODE that has no effect (no transaction is present).